File: inputitem.cpp
#include "inputitem.h"

#include "retta.h"
#include "poligono.h"


inputitem::~inputitem() {}

inputitem* inputitem::iniz_input (string st) {

    if(dynamic_cast<retta*>(inputitem::pars_start(st))){
        retta ret;
        try{ret.pars_rect(st);}
        catch(int){
            return new retta(ret);
        }
        catch(...){
            throw;
        }
    }
    else if(dynamic_cast<punto*>(inputitem::pars_start(st)))
    {
        punto point;
        try{point.pars_point(st);}
        catch(input_error){
            throw;
        }
        catch(num_error){
            throw;
        }
        catch(den_error){
            throw;
        }
        catch(int){
            return new punto(point);
        }
    }
    else{
        poligono* p;
        bool errore = true;
        try{ p = poligono::pars_pol(st); }
        catch( input_error ){
            throw;
        }
        catch( irregular_pol ){
            throw;
        }
        return p;
    }
}



//data una string ritorna se è un punto , retta o poligono
//ha senso creare un oggetto e ritornarlo  esfruttare il polimorfisomo? altrimenti stringa
inputitem* inputitem::pars_start(string s){
    unsigned int len = s.length();

    for (unsigned int var = 0; var < len; ++var) {
        if(s[var] == ' '){
            s.erase(s.begin()+var);
            var--;
        }
    }
    //tolto gli spazi

    if(s[0] == '('){
        int cont = 0;
        for (unsigned int i = 0; i < len; ++i) {
            if(s[i] == '('){
                cont++;
            }
        }
        if( cont > 1 ) return 0;
        else return new punto();
    }
    else{
        return new retta();
    }
}

vector<punto> inputitem::intersect(inputitem * i1 , inputitem * i2 ){
    vector<punto> p ;

    if(dynamic_cast<punto*>(i1) || dynamic_cast<punto*>(i2)){
        //c'è almeno un punto
        if(dynamic_cast<retta*>(i1) || dynamic_cast<retta*>(i2)){
            retta * r1 = dynamic_cast<retta*>(i1);
            punto * p1 = dynamic_cast<punto*>(i2);

            if( !r1 ){
                retta * r1 = dynamic_cast<retta*>(i2);
                punto * p1 = dynamic_cast<punto*>(i1);
                double dist = retta::distancePuntoRetta(*p1,*r1);
                if( dist == 0 ) {
                    p.push_back(*p1);
                }
            }else{
                double dist = retta::distancePuntoRetta(*p1,*r1);
                if( dist == 0 ) {
                    p.push_back(*p1);
                }
            }
            return p;
        }
        else if(dynamic_cast<poligono*>(i1) || dynamic_cast<poligono*>(i2)){
            //intersezione poligono punto

        }
        else {
            //sono due punti
            punto * p1 = dynamic_cast<punto*>(i1);
            punto * p2 = dynamic_cast<punto*>(i2);

            if(punto::distanceTwoPoints(*p1,*p2) == 0){
                p.push_back(*p1);
            }
            return p;
        }
    }
    else {
        //non ci sono punti
        if(dynamic_cast<retta*>(i1) || dynamic_cast<retta*>(i2)){
            if(dynamic_cast<poligono*>(i1) || (dynamic_cast<poligono*>(i2))){
                retta * r1 = dynamic_cast<retta*>(i1);
                poligono * pol1 = dynamic_cast<poligono*>(i2);
                if( !r1 ){
                    retta * r1 = dynamic_cast<retta*>(i2);
                    poligono * pol1 = dynamic_cast<poligono*>(i1);
                    return pol1->rettapol(r1);
                }

                return pol1->rettapol(r1);
            }
            else{
                //invoco retta retta
                retta * r1 = dynamic_cast<retta*>(i1);
                retta * r2 = dynamic_cast<retta*>(i2);
                return retta::Intersect(*r1,*r2);
            }
        }
        else {
            poligono * pol1 = dynamic_cast<poligono*>(i1);
            poligono * pol2 = dynamic_cast<poligono*>(i2);
            return pol1->polipoli(pol2);
        }
    }

}

File: java/inputitem.java
import java.util.Vector;

public abstract class inputitem{
	
	public static inputitem pars_start(String b){
		String input[] = new String[b.length()];
	    input = b.split("");

	    //parsing iniziale per verifica la figura.
	    //- se ci sono n. "(" > 1 è chiaramente un poligono
	    //- se è presente una volta "(" allora è un punto
	    //- altrimenti è una retta

	    int len = input.length;
		if(input[0].equals("(")){
	        int cont = 0;
	        for (int i = 0; i < len; ++i) {
	            if(input[i].equals("(")){
	                cont++;
	            }
	        }
	        if( cont > 1 ) return new quadrato();
	        else return new punto();
	    }
	    else{
	        return new retta();
	    }
	}

	public inputitem iniz_input(String b) throws eccezioni{
		if(pars_start(b) instanceof retta){
			retta r = new retta();
			try{r.pars_rect(b);}
			catch(eccezioni e){
				System.out.println(e.message);
				throw e;
			}

			return r;
		}
		else if(pars_start(b) instanceof punto){
			punto p = new punto();

			try{p.pars_point(b);}
			catch(eccezioni e){
				System.out.println(e.message);
				throw e;
			}

			return p;
		}
		else{
			poligono p = new quadrato();
			try{p = p.pars_pol(b);}
			catch(eccezioni e){
				System.out.println(e.message);
				throw e;
			}
			return p;
		}


	}

	public abstract double distance(inputitem i) throws eccezioni;

	public abstract Vector<punto> intersect(inputitem i);

}